src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/HeartbeatProvider.cs

```diff
@@ -0,0 +1,312 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Globalization;
+    using System.Linq;
+    using System.Threading;
+    using Microsoft.ApplicationInsights.Channel;
+    using Microsoft.ApplicationInsights.DataContracts;
+    using Microsoft.ApplicationInsights.Extensibility;
+
+    /// <summary>
+    /// Implementation of health heartbeat functionality.
+    /// </summary>
+    internal class HeartbeatProvider : IDisposable, IHeartbeatProvider
+    {
+        /// <summary>
+        /// The default interval between heartbeats if not specified by the user. Left public for use in unit tests.
+        /// </summary>
+        public static readonly TimeSpan DefaultHeartbeatInterval = TimeSpan.FromMinutes(15.0);
+
+        /// <summary>
+        /// The name of the health heartbeat metric item and operation context. 
+        /// </summary>
+        private static string heartbeatSyntheticMetricName = "HealthState";
+
+        private readonly List<string> disabledDefaultFields = new List<string>(); // string containing fields that are not to be sent with the payload. Empty list means send everything available.
+
+        private UInt64 heartbeatsSent; // counter of all heartbeats
+
+        /// <summary>
+        /// The payload items to send out with each health heartbeat.
+        /// </summary>
+        private ConcurrentDictionary<string, HeartbeatPropertyPayload> heartbeatProperties;
+
+        private bool disposedValue = false; // To detect redundant calls to dispose
+        private TimeSpan interval; // time between heartbeats emitted
+        private TelemetryClient telemetryClient; // client to use in sending our heartbeat
+        private volatile bool isEnabled; // no need for locks or volatile here, we can skip/add a beat if the module is disabled between heartbeats
+
+        public HeartbeatProvider()
+        {
+            this.interval = DefaultHeartbeatInterval;
+            this.heartbeatProperties = new ConcurrentDictionary<string, HeartbeatPropertyPayload>(StringComparer.OrdinalIgnoreCase);
+            this.heartbeatsSent = 0; // count up from construction time
+            this.isEnabled = true;
+        }
+
+        /// <summary>
+        /// Gets or sets the currently defined interval between heartbeats
+        /// </summary>
+        public TimeSpan HeartbeatInterval
+        {
+            get => this.interval;
+            set
+            {
+                if (value == null || value.TotalMilliseconds <= 0)
```

Comment:
question - what if the value is very low? I mean if this value is much lower than the default FlushInterval for the channel?

Reply:
Very good, I like that much better. We should ensure that this value isn't lower than the default FlushInterval - I will look up how to get that  value and set it to that as the minimum instead.

Reply:
Actually, now that I look at it we can have multiple Channels if I am not mistaken. How would I determine the value of the sending interval with any accuracy? Perhaps it would be better here, as the 'HeartbeatProvider' to select some reasonable default and never drop below it... maybe something like 30 seconds... 

I'll set this default and we can follow up with an issue after the feature is in place.

Reply:
Yes you are right.. I am not sure on how to find the flush interval for the channel reliably.
For now, maybe modify the comment about hbinterval to indicate that - 'setting hb interval to very low values may produce hearbeats at that interval, but need not be send to AI '..

I would like to avoid customer confusion here - someone setting hearbeat interval to 2msec might expect hb every 2msec in portal, which won't be the case if channel does not flush every 2msec.